ninja.lua

```diff
@@ -395,109 +395,82 @@ local function c_cpp_compilation_rules(cfg, toolset, pch)
 	local all_ldflags = getldflags(toolset, cfg)
 	local all_resflags = getresflags(toolset, cfg, cfg)
 
-	if toolset == p.tools.msc then
-		p.outln("CFLAGS=" .. all_cflags)
-		p.outln("rule cc")
-		p.outln("  command = " .. cc .. " $CFLAGS" .. " /nologo /showIncludes -c /Tc$in /Fo$out")
-		p.outln("  description = cc $out")
-		p.outln("  deps = msvc")
-		p.outln("")
-		p.outln("CXXFLAGS=" .. all_cxxflags)
-		p.outln("rule cxx")
-		p.outln("  command = " .. cxx .. " $CXXFLAGS" .. " /nologo /showIncludes -c /Tp$in /Fo$out")
-		p.outln("  description = cxx $out")
-		p.outln("  deps = msvc")
-		p.outln("")
-		p.outln("CFLAGS=" .. all_cflags)
-		p.outln("rule clangtidy_cc")
-		p.outln("  command = clang-tidy $in -- -x c $CFLAGS &&$")
-		p.outln("            " .. cc .. " $CFLAGS" .. " /nologo /showIncludes -c /Tc$in /Fo$out")
-		p.outln("  description = cc $out")
-		p.outln("  deps = msvc")
-		p.outln("")
-		p.outln("CXXFLAGS=" .. all_cxxflags)
-		p.outln("rule clangtidy_cxx")
-		p.outln("  command = clang-tidy $in -- -x c++ $CFLAGS &&$")
-		p.outln("            " .. cxx .. " $CXXFLAGS" .. " /nologo /showIncludes -c /Tp$in /Fo$out")
-		p.outln("  description = cxx $out")
-		p.outln("  deps = msvc")
-		p.outln("")
-		p.outln("RESFLAGS = " .. all_resflags)
-		p.outln("rule rc")
-		p.outln("  command = " .. rc .. " /nologo /fo$out $in $RESFLAGS")
-		p.outln("  description = rc $out")
+	local function emit_rule(name, cmd_parts, description, opts)
+		opts = opts or {}
+		p.outln("rule " .. name)
+		p.outln("  command = " .. table.concat(cmd_parts, " "))
+		p.outln("  description = " .. description)
+		if opts.deps then
+			p.outln("  deps = " .. opts.deps)
+		end
+		if opts.depfile then
+			p.outln("  depfile = " .. opts.depfile)
+		end
 		p.outln("")
+	end
+
+	local function emit_flags(name, value)
+		p.outln(name .. "=" .. value)
+	end
+
+	local inp, outp = "$in", "$out"
```

Comment:
`p` suffix seems strange. I would prefer something like `in_variable`/`in_placeholder`.
I think those names might also match the future relative case.